### ISSUE TITLE
Coordinator blocks on storage future for an unexpected amount of time

### ISSUE THREAD

**Issue by jpepin** [MaterializeInc/materialize#26172]
### What version of Materialize are you using?

v0.90.1, v0.91.0, v0.92.1

### What is the issue?

Running a [version](https://github.com/MaterializeInc/cloud/pull/8884) of the prometheus sql exporter modified to retry collection when a replica times out can cause environmentd to become unresponsive for up to several hours due to the coordinator blocking on a storage future.
So far this has only been reproduced in a personal cloud stack.

A synopsis of the details gathered so far:
>The time there is spent in group_commit_apply::append_fut which is awaiting the future returned from a call to storage.append_table.  This is in a blocking path in the coordinator, and so makes that env unresponsive

[slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1710306702498689)

```
     Image:         materialize/environmentd:v0.92.1                                                                                                               │
     Image ID:      docker.io/materialize/environmentd@sha256:3277ba1ef353f1fd87d584c2ca32f2bc0787d84a38786d768124d67c97c8444a
```
This occurs in a personal stack environment which has in many cases smaller resource requests for mz components, and could experience LD drift from what is set in staging and prod.

Relates to https://github.com/MaterializeInc/accounts/issues/58

### Traces
[Trace-1e712d-2024-03-21 14 25 32.json](https://github.com/MaterializeInc/materialize/files/14696650/Trace-1e712d-2024-03-21.14.25.32.json)
[Trace-00548a-2024-03-21 14 26 24.json](https://github.com/MaterializeInc/materialize/files/14696651/Trace-00548a-2024-03-21.14.26.24.json)
[Trace-e48528-2024-03-21 14 21 10.json](https://github.com/MaterializeInc/materialize/files/14696657/Trace-e48528-2024-03-21.14.21.10.json)

**Comment by jkosh44** [comment]
> The time there is spent in group_commit_apply::append_fut which is awaiting the future returned from a call to storage.append_table. This is in a blocking path in the coordinator, and so makes that env unresponsive

I haven't read the full slack thread, but does this come from a trace? `group_commit_apply` does not call `storage.append_table`.

https://github.com/MaterializeInc/materialize/blob/aa95468609d557cd09619073aa54e49f4db89760/src/adapter/src/coord/appends.rs#L432-L456

`group_commit_initiate` does call `storage.append_table`, which produces a future called `append_fut`, but that future is awaited in a separate task off of the blocking path in the coordinator.

https://github.com/MaterializeInc/materialize/blob/aa95468609d557cd09619073aa54e49f4db89760/src/adapter/src/coord/appends.rs#L361-L418

So I think something else must be going on.

**Comment by jpepin** [comment]
> I haven't read the full slack thread, but does this come from a trace? `group_commit_apply` does not call `storage.append_table`.

It does come from a trace, which unfortunately we have lost to a Tempo TTL. I am working to reproduce and capture another trace, which I'll attach here

**Comment by jpepin** [comment]
Traces attached to the issue! You can import these to Grafana for viewing

**Comment by jkosh44** [comment]
> Traces attached to the issue! You can import these to Grafana for viewing

Awesome! All these traces pretty much have the same format, so I'll just talk about [Trace-e48528-2024-03-21 14 21 10.json](https://github.com/MaterializeInc/materialize/files/14696657/Trace-e48528-2024-03-21.14.21.10.json). They all correspond to a storage usage update. What's happening is that the Coordinator main loop is stuck waiting for the `storage.append` call to complete. The trace is a little hard to follow, for example `coord::group_commit_initiate` takes 97.47 ms, while `group_commit_apply`, which is under `coord::group_commit_initiate`, takes `1h 26m`.

![Screenshot from 2024-03-21 14-31-23](https://github.com/MaterializeInc/materialize/assets/10800303/75727466-837f-49bd-b03f-1e1806991b3c)

What's happening is the following (all in `mz_adapter::coord::message_handler::storage_usage_update`):

1. Get a timestamp from the timestamp oracle (46.59 ms)
2. Update the catalog (456.61 ms)
3. Trigger a group commit (97.47 ms)
  a. Create a new channel.
  b. Get a future from `storage.append` (`append_fut`).
  c. Await the future on another task.
  d. Send a message to the channel in the other task to indicate that the append has completed.
5. Wait for a message in the main coord thread to let us know that the append has completed (1h 26m).

I can't remember why the main coord thread needs to wait for the appends to complete for storage usage updates. If possible we should just not wait on the main coord thread.

**Comment by jkosh44** [comment]
> I can't remember why the main coord thread needs to wait for the appends to complete for storage usage updates. If possible we should just not wait on the main coord thread.

Looking at the code I think the only reason it blocks is because it borrows the same code path as DDL and DDL does need to block. Storage usage is not really a DDL, but it does update the catalog. I think we should should be able to remove the blocking without any correctness issues. I'll work on this now.

**Comment by jkosh44** [comment]
Not blocking on the call to `append` should resolve this issue, however it's still sounds pretty scary that a call to `append` can take 1.5 hours. @MaterializeInc/storage @MaterializeInc/persist any thoughts on this?

**Comment by jpepin** [comment]
> Not blocking on the call to `append` should resolve this issue, however it's still sounds pretty scary that a call to `append` can take 1.5 hours.

It's worth noting that this is occurring in a personal stack which has drastically lower resource requests and sizes for almost every component (envd, crdb, nodes) than staging or production Materialize, so this may be replicating an extremely resource starved behavior. There could also be differences in the LD flag settings. We haven't attempted to repro this elsewhere yet.

**Comment by danhhz** [comment]
> Not blocking on the call to `append` should resolve this issue, however it's still sounds pretty scary that a call to `append` can take 1.5 hours. @MaterializeInc/storage @MaterializeInc/persist any thoughts on this?

Certainly worrisome, but hard to say much without knowing more about how those 1.5 hours are being spent

**Comment by danhhz** [comment]
Unfortunately, it doesn't look like our tracing spans have this `append` call hooked up :(

<img width="1803" alt="image" src="https://github.com/MaterializeInc/materialize/assets/52528/50234672-6131-4d7c-971c-2c0e054ab126">

**Comment by jkosh44** [comment]
@jpepin, this branch should fix your issue if it's possible to re-run your tests on a different branch: https://github.com/MaterializeInc/materialize/pull/26193

If that fixes your issues, then I'd also be curious to see what happens if you try running a bunch of DDL.

**Comment by jpepin** [comment]
> @jpepin, this branch should fix your issue if it's possible to re-run your tests on a different branch: #26193

Brought this up on my personal stack and it looks good. I haven't seen any more stuck coord messages on this version.

> If that fixes your issues, then I'd also be curious to see what happens if you try running a bunch of DDL.

How much DDL is a bunch? Would a CREATE and a DROP TABLE in a loop for a few hours do it?

**Comment by jkosh44** [comment]
> How much DDL is a bunch? Would a CREATE and a DROP TABLE in a loop for a few hours do it?

Yeah I was thinking around 10-20 DDL. I just wanted to see if any blocking calls to `append` would get stuck or if it was specifically the storage usage collection calls to `append`.

**Comment by jkosh44** [comment]
Something interesting about this that updating the catalog is always fast (< 1s), but a call to `append` is always slow. However, under the hood they both are just a single write to persist. So why do we see such a drastic difference in latency? If this is truly just a result of CRDB running hot, then we'd expect to also see the catalog update take a long time.

**Comment by jkosh44** [comment]
> Something interesting about this that updating the catalog is always fast (< 1s), but a call to `append` is always slow. However, under the hood they both are just a single write to persist. So why do we see such a drastic difference in latency? If this is truly just a result of CRDB running hot, then we'd expect to also see the catalog update take a long time.

I just realized that the big difference is that `append` will go through persist txns and writes to all tables, while the catalog write directly writes to a single shard.

**Comment by jkosh44** [comment]
@benesch @jpepin do we want to re-open this until we figure out why `append` takes 1.5 hrs?

**Comment by jpepin** [comment]
> @benesch @jpepin do we want to re-open this until we figure out why `append` takes 1.5 hrs?

I'm happy to keep investigating this! However, as @pH14 and others have pointed out, personal stacks have so many differences from staging and prod that I don't want to overindex on failures that we're seeing there.
A staging environment under the same conditions has not seen the same level of slow coordinator messages (though [envd, the system and introspection clusters](https://grafana.dev.materialize.com/d/UT2whmGVk/kube-pod-metrics?orgId=1&var-datasource=Ks85Oh14z&var-namespace=environment-2d677fbc-16c8-4263-9359-908a47140dc2-0&var-pod=All&var-container=All&from=now-2d&to=now&viewPanel=4), and [update_state_slow_path](https://grafana.dev.materialize.com/d/bf6bcae5-1a21-4649-9386-8eaaf2745d64/jp-env-health-scratch?orgId=1&refresh=10s&var-version=All&var-datasource=Ks85Oh14z&var-organization=2d677fbc-16c8-4263-9359-908a47140dc2&var-mz_cluster=All&var-stack=staging&var-Org_Lookup_not_a_filter=pstoev-loadtest-tenant6%20-%20dd4b7ea2-e9fc-4880-a03a-6a5f44342310&var-e2e_tenants=06ee07ee-e4f0-daff-f992-ac841534175a%7C07ca2aef-58d3-f581-ee36-a1af3949ef6f%7C15646b33-f169-3ba0-e480-949ab67dbd4e%7C1f6dc656-23b9-d8bf-d471-10ac9024b383%7C202f9221-2d17-a450-aa40-98d2055ca74b%7C20fb86cc-400b-fb95-cddc-2247b2bd63ec%7C647104f4-c8b7-6df4-b6c4-a9381ed23624%7C66a7b87a-8bd0-b892-945f-e84d3cb68dfd%7C679b8ba7-8bbb-a78a-18ca-ef03c89fe6c3%7C7184197a-67ae-0dbe-ca9d-f502a3223823%7C7d246218-cb7d-f2d2-0896-bc8ccf4f6562%7C9505cba4-f80b-5cb4-3dd9-fb661222dcdc%7C9e754f72-cf6f-7be4-4982-cd2275fa1d8e%7Ca9f6623e-b7e4-6a32-fffa-fd6a7106029d%7Cad0d3a5a-bbf8-05ee-c26d-254fd3d9e818%7Cb68f108c-116e-3ca9-af57-9a5f8de86fc6%7Cb80da966-401a-993e-fa81-182971de501f%7Cb8f06143-2b62-b96e-7ba0-c39aeb9d758f%7Cbaec7d0f-6315-ae19-e2f3-215ea8e2eae6%7Cf635ade6-4dd1-b103-0d21-7e35c5d010b2%7C06ec6d64-e4c7-fc62-9477-fbfd06c401cd%7C495bf6ad-5ff6-34e8-cba1-a811e7154b77%7C62383511-4d0f-b115-664f-9a820501dca8%7Cbd4e9d69-246d-6fb8-c251-63e162b03767%7Cd9476957-6d7f-ea94-0812-9d8195382435%7C1d00e33b-7e65-9376-7b68-2d98cb400c14%7C2184850f-6d48-9d88-c0a6-a1ffee4b48dc%7C30ef8e02-65bd-d1c3-6ba8-d6ac583e48d9%7C4d6615e8-7c23-bcf2-2376-51bac413b010%7C575f7ee8-7a95-652a-9c49-7c12ada4938e%7C6a0b2af5-eda8-4d47-f49e-01836e0b05ba%7C6a4346a4-b2e0-00f3-293f-2eb0ef1b09ae%7C84555dc6-188f-dcd2-2b91-b46d7a76f8de%7C95f5ed65-a73f-c323-e406-b47b8098ce90%7Cb94241a3-8f30-3ab6-5813-f1c68f701164%7Cbbdaf0d8-7532-c9f7-e143-fd5f61b1658c%7Cc082bbd1-f643-f975-fa3b-f7f5a7413c2c%7Cc45a3f38-4ebb-0c24-6fe8-55d0f49f1e5a%7Cc5c27d9a-ef61-e639-070d-04f188819974%7Cce70efd2-bc21-b3a6-5b78-676c4791d672%7Cd7864e16-8485-d572-d219-c55041915b82%7Cd85e8542-94a5-25c9-437e-158613ff9de7%7Cdca5641d-4d89-fae7-17bd-cfda5bef62e1%7Cdd7f89a9-5a89-a927-2e73-bec4d7701d0a%7Cff770cfd-28f6-2192-034d-541e489e5e18%7C0512204c-ae47-16ae-4691-6422be28c508%7C60a595b4-99a9-bdc5-49b7-308b3624049d%7C982a2d9c-8db9-3da4-8a76-142770552f6e%7Cb7e60d7a-e8f6-0bea-5ead-cc9466a375bd%7Cc8b4ba99-7efd-aa77-43be-bf9798e904a1&from=now-2d&to=now&viewPanel=8) are working harder than before).

If we think it's useful to understand more deeply as a type of failure scenario that occurs when crdb becomes overloaded, I'm happy to keep working on it.

**Comment by benesch** [comment]
It's definitely scary, but I'm comfortable chalking this up to dev CockroachDB overload for now. Are we able to increase the size of the dev CockroachDB cluster to get you unblocked?

**Comment by jpepin** [comment]
> Are we able to increase the size of the dev CockroachDB cluster to get you unblocked?

Yes, I can do this today
